python/docs/source/user_guide/python_packaging.rst

```diff
@@ -0,0 +1,201 @@
+
+..    http://www.apache.org/licenses/LICENSE-2.0
+
+..  Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+=========================
+3rd Party Python Packages
+=========================
+
+When you want to run your PySpark application on a cluster such as YARN, Kubernetes, Mesos, etc., you need to make
+sure that the your code and all used libraries are available on the executors.
+
+As an example let's say you may want to run the `Pandas UDF's examples <arrow_pandas.rst#series-to-scalar>`_.
+As it uses pyarrow as an underlying implementation we need to make sure to have pyarrow installed on each executor
+on the cluster. Otherwise you may get errors such as ``ModuleNotFoundError: No module named 'pyarrow'``.
+
+Here is the script ``app.py`` from the previous example that will be executed on the cluster:
+
+.. code-block:: python
+
+    import pandas as pd
+    from pyspark.sql.functions import pandas_udf
+    from pyspark.sql import SparkSession
+
+    def main(spark):
+        df = spark.createDataFrame(
+            [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
+            ("id", "v"))
+
+        @pandas_udf("double")
+        def mean_udf(v: pd.Series) -> float:
+            return v.mean()
+
+        print(df.groupby("id").agg(mean_udf(df['v'])).collect())
+
+
+    if __name__ == "__main__":
+        main(SparkSession.builder.getOrCreate())
+
+
+There are multiple ways to ship the dependencies to the cluster:
+
+- Using PySpark Native Features
+- Using Zipped Virtual Environment
+- Using PEX
+
+
+Using PySpark Native Features
+-----------------------------
+
+PySpark allows to upload Python files (``.py``), zipped Python packages (``.zip``), and Egg files (``.egg``)
+to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly
+calling :meth:`pyspark.SparkContext.addPyFile`.
+
+This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole
+packages and upload them. Using :meth:`pyspark.SparkContext.addPyFile` allows to upload code
+even after having started your job.
+
+Note that it doesn't allow to add packages built as `Wheels <https://www.python.org/dev/peps/pep-0427/>`_ and therefore doesn't
+allowing to include dependencies with native code.
+
+
+Using Zipped Virtual Environment
+--------------------------------
+
+The idea of zipped environments is to zip your whole `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_, 
+ship it to the cluster, unzip it remotly and target the Python interpreter from inside this zipped environment. Note that this
+is currently supported *only for YARN*.
+
+Zip Virtual Environment
```

Comment:
Could we emphasize here, that these solutions work only on YARN?

Reply:
Actually I did above `*only for YARN*` .. :-). If that sounds not enough, I am fine by emphasizing more.

Reply:
Sorry, I am thinking more about putting this note in the header or something equivalent :)

Reply:
Putting besides Kubernetes which is all docker based I don't see a restriction on why this couldn't work with other deployements modes. 
I know Mesos pretty well as we also have to a mesos cluster (even though I have never tried to deploy Sprak on it) and I think it will work the same. For Standalone mode I can confirm it works.

So spark-submit I'm pretty sure it works. The restriction to me is only for running SparkSession.Builder from inside Python where there is one argument missing to inject archives as config. But it is a minor fix imo. I could even do it. I would prefer not saying restricted to YARN and do the fix instead and inject archives via a config param.

Reply:
In such case could we add (and validate) non-YARN example? That would be useful, especially when both `conda-pack` and   `venv-pack` already provide YARN ones.

Reply:
Sorry for the confusion. I had a look again and the feature that zip files are extracted on the executor really seems specific to YARN (https://github.com/apache/spark/blob/3edab6cc1d70c102093e973a2cf97208db19be8c/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala#L563. I also tried and  '--archives' actually does nothing in Standalone mode.
So the current state is good. Zippable env only supported for YARN. PEX should work everywhere as it is shipped as a file.
So I guess